2011-06-05  Michael Tautschnig  <mt@debian.org>

	* setup-storage: added option -s to perform syntax check only
		(closes: #627913)
	* setup-storage/{Init.pm,Parser.pm}: ignore several checks if doing syntax
		check only.

Index: trunk/bin/setup-storage
===================================================================
--- trunk.orig/bin/setup-storage
+++ trunk/bin/setup-storage
@@ -59,14 +59,15 @@
 use Getopt::Std;
 $main::VERSION = $version;
 $Getopt::Std::STANDARD_HELP_VERSION = 1;
-our ($opt_X, $opt_f, $opt_h, $opt_d); # the variables for getopt
-(&getopts('Xf:hd') && !$opt_h) || die <<EOF;
+our ($opt_X, $opt_f, $opt_h, $opt_d, $opt_s); # the variables for getopt
+(&getopts('Xf:hds') && !$opt_h) || die <<EOF;
 setup-storage version $version
 
 USAGE: [-X]                     no test, your harddisks will be formated
                                 default: only test, no real formating
        [-f<config-filename>]    default: parse classes
        [-d]                     enable debug output (equivalent to debug=1)
+       [-s]                     perform syntax check only and exit
        [-h]                     print this help message
 EOF
 
@@ -87,6 +88,9 @@
 $opt_X and $FAI::no_dry_run = 1;
 warn "setup-storage is running in test-only mode\n" unless ($FAI::no_dry_run);
 
+# syntactic checks only
+$opt_s and $FAI::check_only = 1;
+
 # Find out whether $LOGDIR is usable or default to /tmp/fai
 if (! -d $FAI::DATADIR) {
   defined ($ENV{LOGDIR}) and die
@@ -137,6 +141,12 @@
 # make sure there are no empty disk_config stanzas
 &FAI::check_config;
 
+if ($FAI::check_only)
+{
+  print "Syntax ok\n";
+  exit 0;
+}
+
 # first find the proper way to tell udev to settle
 $FAI::udev_settle = "udevadm settle --timeout=10" if (&FAI::in_path("udevadm"));
 $FAI::udev_settle = "udevsettle --timeout=10" if (&FAI::in_path("udevsettle"));
Index: trunk/lib/setup-storage/Init.pm
===================================================================
--- trunk.orig/lib/setup-storage/Init.pm
+++ trunk/lib/setup-storage/Init.pm
@@ -64,6 +64,13 @@
 
 ################################################################################
 #
+# @brief Perform syntactic checks only if set to 1
+#
+################################################################################
+$FAI::check_only = 0;
+
+################################################################################
+#
 # @brief The command to tell udev to settle (udevsettle or udevadm settle)
 #
 ################################################################################
Index: trunk/lib/setup-storage/Parser.pm
===================================================================
--- trunk.orig/lib/setup-storage/Parser.pm
+++ trunk/lib/setup-storage/Parser.pm
@@ -61,6 +61,9 @@
   # initialize the parameter
   my ($cmd) = @_;
 
+  # ignored in syntax-check mode
+  return 1 if ($FAI::check_only);
+
   # split $PATH into its components, search all of its components
   # and test for $cmd being executable
   (-x "$_/$cmd") and return 1 foreach (split (":", $ENV{PATH}));
@@ -82,9 +85,11 @@
 sub resolve_disk_shortname {
   my ($disk) = @_;
 
+  $disk = "sdx" . chr(ord('a') + $disk - 1)
+    if ($FAI::check_only && $disk =~ /^\d+$/);
+
   # test $disk for being numeric
   if ($disk =~ /^\d+$/) {
-
     # $disk-1 must be a valid index in the map of all disks in the system
     (scalar(@FAI::disks) >= $disk)
       or die "this system does not have a physical disk $disk\n";
